bundle_install_process.go

```diff
@@ -15,69 +18,120 @@ type Executable interface {
-	buffer := bytes.NewBuffer(nil)
-	args := []string{"config", "path", gemLayersDir}
-
-	ip.logger.Subprocess("Running 'bundle %s'", strings.Join(args, " "))
-	err := ip.executable.Execute(pexec.Execution{
-		Args:   args,
-		Stdout: buffer,
-		Stderr: buffer,
-	})
+func (ip BundleInstallProcess) ShouldRun(layer packit.Layer, workingDir string) (bool, string, string, error) {
+	rubyVersion, err := ip.versionResolver.Lookup()
 	if err != nil {
-		return fmt.Errorf("failed to execute bundle config output:\n%s\nerror: %s", buffer.String(), err)
+		return false, "", "", err
 	}
 
-	buffer = bytes.NewBuffer(nil)
-	args = []string{"config", "without", "development:test"}
+	cachedRubyVersion, ok := layer.Metadata["ruby_version"].(string)
+	rubyVersionMatch := true
 
-	ip.logger.Subprocess("Running 'bundle %s'", strings.Join(args, " "))
-	err = ip.executable.Execute(pexec.Execution{
-		Args:   args,
-		Stdout: buffer,
-		Stderr: buffer,
-	})
+	if ok {
+		rubyVersionMatch, err = ip.versionResolver.CompareMajorMinor(cachedRubyVersion, rubyVersion)
+		if err != nil {
+			return false, "", "", err
+		}
+	}
+
+	var sum string
+	_, err = os.Stat(filepath.Join(workingDir, "Gemfile.lock"))
 	if err != nil {
-		return fmt.Errorf("failed to execute bundle config output:\n%s\nerror: %s", buffer.String(), err)
+		if !os.IsNotExist(err) {
+			return false, "", "", err
+		}
+	} else {
+		sum, err = ip.calculator.Sum(filepath.Join(workingDir, "Gemfile"), filepath.Join(workingDir, "Gemfile.lock"))
+		if err != nil {
+			return false, "", "", err
+		}
 	}
 
-	buffer = bytes.NewBuffer(nil)
-	args = []string{"config", "clean", "true"}
+	cachedSHA, ok := layer.Metadata["cache_sha"].(string)
+	cacheMatch := ok && cachedSHA == sum
+	shouldRun := !cacheMatch || !rubyVersionMatch
 
-	ip.logger.Subprocess("Running 'bundle %s'", strings.Join(args, " "))
-	err = ip.executable.Execute(pexec.Execution{
-		Args:   args,
-		Stdout: buffer,
-		Stderr: buffer,
-	})
+	return shouldRun, sum, rubyVersion, nil
+}
 
-	if err != nil {
-		return fmt.Errorf("failed to execute bundle config output:\n%s\nerror: %s", buffer.String(), err)
+func (ip BundleInstallProcess) Execute(workingDir, layerPath string, config map[string]string) error {
+	localConfigPath := filepath.Join(workingDir, ".bundle", "config")
+	globalConfigPath := filepath.Join(layerPath, "config")
+	if _, err := os.Stat(localConfigPath); err == nil {
+		err := os.MkdirAll(layerPath, os.ModePerm)
+		if err != nil {
+			return err
+		}
+
+		err = fs.Copy(localConfigPath, globalConfigPath)
+		if err != nil {
+			return err
+		}
 	}
 
-	buffer = bytes.NewBuffer(nil)
-	args = []string{"config", "cache_path", "--parseable"}
+	env := append(os.Environ(), fmt.Sprintf("BUNDLE_USER_CONFIG=%s", globalConfigPath))
+
+	var keys []string
+	for key := range config {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	for _, key := range keys {
+		buffer := bytes.NewBuffer(nil)
+		args := []string{"config", "--global", key, config[key]}
```

Comment:
Referring back to my question above, why are you opting for `--global` rather than `--local`?

Reply:
I'm using `--global` to make sure we are modifying the config that is set by `BUNDLE_USER_CONFIG` instead of `--local` which would modify the `.bundle/config` that is local to the `/workspace` directory.